gms/gossiper.hh

```diff
@@ -434,6 +434,16 @@ private:
      */
     future<> handle_major_state_change(inet_address ep, const endpoint_state& eps);
 
+    template <typename Func>
+    auto run_on_shard0_with_endpoint_update_semaphore(Func&& func) {
+        return container().invoke_on(0, [this, func = std::forward<Func>(func)] (gms::gossiper& g) {
```

Comment:
need to specify `mutable` here so `func` can be moved below.
Also, if you `co_await container().invoke_on`, you could use a lambda coroutine and `co_await get_units`, etc. simplifying the whole thing (I think).